File: 07-Spam Classification/ex7.py
import re
from nltk.stem.snowball import SnowballStemmer
import numpy as np
from scipy import io
from sklearn import svm

def get_vocablist():
    vocabulary = []
    with open('vocab.txt') as f:
        for line in f:
            idx, word = line.split('\t')
            vocabulary.append(word.strip())
    return vocabulary

def split(delimiters, string, maxsplit=0):
    pattern = '|'.join(map(re.escape, delimiters))
    return re.split(pattern, string, maxsplit)

def process_email(email_contents):
    vocab_list = get_vocablist()
    email_contents = email_contents.lower()
    email_contents = re.sub('<[^<>]+>', ' ', email_contents)
    email_contents = re.sub('[0-9]+', 'number', email_contents)
    email_contents = re.sub('(http|https)://[^\s]*', 'httpaddr', email_contents)
    email_contents = re.sub('[^\s]+@[^\s]+', 'emailaddr', email_contents)
    email_contents = re.sub('[$]+', 'dollar', email_contents)
    words = split(""" @$/#.-:&*+=[]?!(){},'">_<;%\n\r""", email_contents)
    word_indices = []
    stemmer = SnowballStemmer("english")
    for word in words:
        word = re.sub('[^a-zA-Z0-9]', '', word)
        if word == '':
            continue
        word = stemmer.stem(word)
        if word in vocab_list:
            idx = vocab_list.index(word)
            word_indices.append(idx)
    return word_indices

def email_features(word_indices):
    n = 1899
    x = np.zeros((n, 1))
    x[word_indices] = 1
    return x

if __name__ == "__main__":

    # ==================== Part 1: Email Preprocessing ====================
    print ('Preprocessing sample email (emailSample1.txt)...')
    file = open('emailSample1.txt', 'r')
    file_contents = file.read().replace('\n', '')

    word_indices = process_email(file_contents)
    print ('Word Indices:', word_indices)

    # ==================== Part 2: Feature Extraction ====================
    print ('Extracting features from sample email (emailSample1.txt)...')
    features = email_features(word_indices)
    print ('Length of feature vector:', len(features))
    print ('Number of non-zero entries:', np.sum(features > 0))

    # =========== Part 3: Train Linear SVM for Spam Classification ========
    # Load the Spam Email dataset
    train_data = io.loadmat('spamTrain.mat')
    X_train = train_data['X']
    y_train = train_data['y'].ravel()

    print ('Training Linear SVM (Spam Classification)...')
    C = 1
    clf = svm.LinearSVC(C=C)
    clf.fit(X_train, y_train)
    y_pred_train = clf.predict(X_train)
    print ('Training Accuracy:', np.mean(y_pred_train == y_train) * 100)

    # =================== Part 4: Test Spam Classification ================
    # Load the test dataset
    test_data = io.loadmat('spamTest.mat')
    X_test = test_data['Xtest']
    y_test = test_data['ytest'].ravel()
    print ('Evaluating the trained Linear SVM on a test set...')
    y_pred_test = clf.predict(X_test)
    print ('Test Accuracy:', np.mean(y_pred_test == y_test) * 100)

    # ================= Part 5: Top Predictors of Spam ====================
    coef = clf.coef_.ravel()
    idx = coef.argsort()[::-1]
    vocab_list = get_vocablist()

    print ('Top predictors of spam:')
    for i in range(15):
        print ("{0:<15s} ({1:f})".format(vocab_list[idx[i]], coef[idx[i]]))

    # =================== Part 6: Try Your Own Emails =====================
    file_name = open('spamSample2.txt', 'r')
    file_contents = file_name.read().replace('\n', '')

    word_indices = process_email(file_contents)
    x = email_features(word_indices)
    p = clf.predict(x.T)
    print ('Processed', file_name, '\nSpam Classification:', p)
    print ('(1 indicates spam, 0 indicates not spam)')

File: 07-Spam Classification/ex7.m
fprintf('\nPreprocessing sample email (emailSample1.txt)\n');

function vocabList = getVocabList()
  fid = fopen('vocab.txt');
  n = 1899;
  vocabList = cell(n, 1);
  for i = 1:n
    fscanf(fid, '%d', 1);
    vocabList{i} = fscanf(fid, '%s', 1);
  endfor
  fclose(fid);
endfunction

function word_indices = processEmail(email_contents)
  vocabList = getVocabList();
  word_indices = [];
  email_contents = lower(email_contents);
  email_contents = regexprep(email_contents, '<[^<>]+>', ' ');
  email_contents = regexprep(email_contents, '[0-9]+', 'number');
  email_contents = regexprep(email_contents, '(http|https)://[^\s]*', 'httpaddr');
  email_contents = regexprep(email_contents, '[^\s]+@[^\s]+', 'emailaddr');
  email_contents = regexprep(email_contents, '[$]+', 'dollar');
  fprintf('\n==== Processed Email ====\n\n');
  l = 0;
  while ~isempty(email_contents)
    [str, email_contents] = strtok(email_contents, [' @$/#.-:&*+=[]?!(){},''">_<;%' char(10) char(13)]);
    str = regexprep(str, '[^a-zA-Z0-9]', '');
    try str = porterStemmer(strtrim(str)); 
    catch str = ''; continue;
    end;
    
    if length(str) < 1
      continue;
    end          
    for i=1:length(vocabList)
	    match(i)=strcmp(vocabList{i}, str);
  		if (match(i)==1)
	  		word_indices = [word_indices; i];
		  else
			  word_indices = [word_indices;[]];
		  endif
	  endfor
    if (l + length(str) + 1) > 78
      fprintf('\n');
      l = 0;
    endif
    fprintf('%s ', str);
    l = l + length(str) + 1;
  endwhile
  fprintf('\n\n=========================\n');
endfunction

function x = emailFeatures(word_indices)
  n = 1899;
  x = zeros(n, 1);
  for i=1:length(word_indices)
	  x(word_indices(i)) = 1;
  endfor
endfunction

function sim = linearKernel(x1, x2)
  x1 = x1(:); x2 = x2(:);
  sim = x1' * x2;
endfunction

function sim = gaussianKernel(x1, x2, sigma)
  x1 = x1(:); x2 = x2(:);
  sim = 0;
  sim = exp(-sum((x1-x2).^2)/(2*(sigma ^ 2)));
endfunction

file_contents = readFile('emailSample1.txt');
word_indices  = processEmail(file_contents);
fprintf('Word Indices: \n');
fprintf(' %d', word_indices);
fprintf('\n\n');

fprintf('Program paused. Press enter to continue.\n');
pause;

fprintf('\nExtracting features from sample email (emailSample1.txt)\n');
features = emailFeatures(word_indices);
fprintf('Length of feature vector: %d\n', length(features));
fprintf('Number of non-zero entries: %d\n', sum(features > 0));

fprintf('Program paused. Press enter to continue.\n');
pause;

load('spamTrain.mat');

fprintf('\nTraining Linear SVM (Spam Classification)\n')
fprintf('(this may take 1 to 2 minutes) ...\n')

C = 0.1;
model = svmTrain(X, y, C, @linearKernel);
p = svmPredict(model, X);
fprintf('Training Accuracy: %f\n', mean(double(p == y)) * 100);

load('spamTest.mat');

fprintf('\nEvaluating the trained Linear SVM on a test set ...\n')

p = svmPredict(model, Xtest);

fprintf('Test Accuracy: %f\n', mean(double(p == ytest)) * 100);
pause;

[weight, idx] = sort(model.w, 'descend');
vocabList = getVocabList();

fprintf('\nTop predictors of spam: \n');
for i = 1:15
    fprintf(' %-15s (%f) \n', vocabList{idx(i)}, weight(i));
end

fprintf('\n\n');
fprintf('\nProgram paused. Press enter to continue.\n');
pause;

filename = 'spamSample1.txt';

file_contents = readFile(filename);
word_indices  = processEmail(file_contents);
x = emailFeatures(word_indices);
p = svmPredict(model, x);

fprintf('\nProcessed %s\n\nSpam Classification: %d\n', filename, p);
fprintf('(1 indicates spam, 0 indicates not spam)\n\n');
